File: api-xmlsignature/csharp/src/util/IdGenerator.cs
using System;
using System.Collections.Generic;
using System.Collections;
using xml_signature.tr.gov.tubitak.uekae.esya.api.xmlsignature.util;

namespace tr.gov.tubitak.uekae.esya.api.xmlsignature.util
{

    using Logger = log4net.ILog;


	/// <summary>
	/// @author ahmety
	/// date: Apr 20, 2009
	/// </summary>
	public class IdGenerator
	{
		private static readonly Logger logger = log4net.LogManager.GetLogger(typeof(IdGenerator));

	    private static readonly IDictionary<string, int> idMap = new SynchronizedDictionary<string, int>();

		public const string TYPE_SIGNATURE = "Signature";
		public const string TYPE_SIGNATURE_VALUE = "Signature-Value";
		public const string TYPE_REFERENCE = "Reference";
		public const string TYPE_SIGNED_INFO = "Signed-Info";
		public const string TYPE_KEY_INFO = "Key-Info";
		public const string TYPE_OBJECT = "Object";
		public const string TYPE_DATA = "Data";
		public const string TYPE_SIGNED_PROPERTIES = "Signed-Properties";


		/// <summary>
		/// Generates sequential ids for different needs. For "Signature" input,
		/// this method generates "Signature-Id-1", "Signature-Id-2",
		/// "Signature-Id-3" etc..
		/// </summary>
		public virtual string uret(string aIdType)
		{
			lock (idMap)
			{
				int index = 0;
				if (idMap.ContainsKey(aIdType))
				{
					index = idMap[aIdType];
				}
				index++;
				idMap[aIdType] = index;
				return aIdType + "-Id-" + index;
			}
		}

		public virtual void update(string id)
		{
			lock (idMap)
			{
				try
				{
					string key = id.Substring(0, id.IndexOf("-Id-"));
					int value = Convert.ToInt32(id.Substring(id.LastIndexOf('-') + 1));

					if (idMap.ContainsKey(key))
					{
						value = Math.Max(idMap[key], value);
					}

					Console.WriteLine("!! update : " + id + "; key: " + key + " : " + value);

					idMap[key] = value;
				}
				catch (Exception exc)
				{
					logger.Debug("Elsewhere generated id : " + id,exc);
					// probably elsewhere generated ids, so no overl,excap should occur
					//x.printStackTrace();
				}
			}
		}
    }

}

File: api-xmlsignature/src/tr/gov/tubitak/uekae/esya/api/xmlsignature/util/IdGenerator.java
package tr.gov.tubitak.uekae.esya.api.xmlsignature.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * @author ahmety
 * date: Apr 20, 2009
 */
public class IdGenerator
{
    private static final Logger logger = LoggerFactory.getLogger(IdGenerator.class);

    private static final Map<String, Integer> idMap = Collections.synchronizedMap(new HashMap<String, Integer>());

    public static final String TYPE_SIGNATURE = "Signature";
    public static final String TYPE_SIGNATURE_VALUE = "Signature-Value";
    public static final String TYPE_REFERENCE = "Reference";
    public static final String TYPE_SIGNED_INFO = "Signed-Info";
    public static final String TYPE_KEY_INFO = "Key-Info";
    public static final String TYPE_OBJECT = "Object";
    public static final String TYPE_DATA = "Data";
    public static final String TYPE_SIGNED_PROPERTIES = "Signed-Properties";


    /**
     * Generates sequential ids for different needs. For "Signature" input,
     * this method generates "Signature-Id-1", "Signature-Id-2",
     * "Signature-Id-3" etc..
     */
    public String uret(String aIdType)
    {
        synchronized (idMap){
            Integer index = 0;
            if (idMap.containsKey(aIdType)){
                index  = idMap.get(aIdType);
            }
            index++;
            idMap.put(aIdType, index);
            return aIdType + "-Id-" + index;
        }
    }

    public void update(String id){
        synchronized (idMap){
            try {
                String key = id.substring(0, id.indexOf("-Id-"));
                Integer value = Integer.parseInt(id.substring(id.lastIndexOf('-')+ 1));

                if (idMap.containsKey(key)){
                    value = Math.max(idMap.get(key), value);
                }

                System.out.println("!! update : "+id+"; key: "+key+" : "+value);

                idMap.put(key, value);
            }
            catch (Exception e){
                logger.debug("Elsewhere generated id : "+id, e);
                // probably elsewhere generated ids, so no overlap should occur
                //x.printStackTrace();
            }
        }
    }

    public static void main(String[] args)
    {
        IdGenerator id = new IdGenerator();
        System.out.println(id.uret(TYPE_SIGNATURE));
        System.out.println(id.uret(TYPE_SIGNATURE));
        System.out.println(id.uret("Signature"));
        System.out.println(id.uret("Signature"));
    }
}
